### PR TITLE
MB-61640: Fuzzy Dynamic Scoring

### PR DESCRIPTION
**Pull request by Likith101**
- Added levenshtein distance calculation for fuzzy and wildcard searchers
 - Added new implementations of certain functions to allow passing of edit distances per term
 - Multiplied boosts by inverse of edit distance for score calculation

### CODE REVIEW

File: search/searcher/search_fuzzy.go [hunk 1]
```diff
@@ -91,9 +92,10 @@ func reportIOStats(ctx context.Context, bytesRead uint64) {
 func findFuzzyCandidateTerms(indexReader index.IndexReader, term string,
 	fuzziness int, field, prefixTerm string) (rv *fuzzyCandidates, err error) {
 	rv = &fuzzyCandidates{
-		candidates: make([]string, 0),
+		candidates:    make([]string, 0),
+		editDistances: make([]uint8, 0),
 	}
-
+	var reuse []int
```
Comment (CascadingRadium) [inline search/searcher/search_fuzzy.go:98]: i think this can be shifted to where it was

Comment (CascadingRadium) [reply search/searcher/search_fuzzy.go:98]: line 139

File: search/searcher/search_fuzzy.go [hunk 2]
```diff
@@ -67,13 +67,14 @@ func NewFuzzySearcher(ctx context.Context, indexReader index.IndexReader, term s
 		}
 	}
 
-	return NewMultiTermSearcher(ctx, indexReader, candidates, field,
-		boost, options, true)
+	return NewMultiTermSearcherBoosted(ctx, indexReader, candidates, field,
+		boost, fuzzyCandidates.editDistances, options, true)
```
Comment (CascadingRadium) [inline search/searcher/search_fuzzy.go:71]: line 54 here indicates that `fuzzyCandidates` can be nil... so i think in line 57.5 you can add editDistances = fuzzyCandidates.editDistances there, just like candidates = fuzzyCandidates.candidates

File: search/searcher/search_multi_term.go [hunk 3]
```diff
@@ -45,6 +45,29 @@ func NewMultiTermSearcher(ctx context.Context, indexReader index.IndexReader, te
 		options, limit)
 }
 
+func NewMultiTermSearcherBoosted(ctx context.Context, indexReader index.IndexReader, terms []string,
```
Comment (CascadingRadium) [inline search/searcher/search_multi_term.go:48]: pls add a comment here regarding what this does

**Review by CascadingRadium — CHANGES_REQUESTED** [review]

**Comment by CascadingRadium** [comment]
also please add unit tests to show the effect (with and w/o boost) across match and match_phrase queries thanks

**Review by CascadingRadium — DISMISSED** [review]

**Review by abhinavdangeti — APPROVED** [review]

**Pull request merged**